File: LinkedList/CircularDoublyLinkList/CDL.java
package DSA.LinkedList.CircularDoublyLinkList;
import java.util.Scanner;

class Node {
    public int data;
    public Node next;
    public Node prev;

    public Node(int data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }
}

public class CDL {
    public Node head;

    public CDL() {
        head = null;
    }

    public void Insert() {
        Scanner sc = new Scanner(System.in);
        System.out.println("\n1.Insert at Start \n2.Middle \n3.End >> ");
        int ch = sc.nextInt();
        System.out.println("Enter the value >> ");
        int data = sc.nextInt();
        Node newNode = new Node(data);

        if (head == null) {
            head = newNode;
            head.next = head;
            head.prev = head;
            return;
        }

        if (ch == 1) {
            newNode.next = head;
            newNode.prev = head.prev;
            head.prev.next = newNode;
            head.prev = newNode;
            head = newNode;
        } else if (ch == 2) {
            Node temp = head;
            System.out.println("After which Node you want to insert data >> ");
            int nextNode = sc.nextInt();

            boolean nodeFound = false;
            do {
                if (temp.data == nextNode) {
                    nodeFound = true;
                    break;
                }
                temp = temp.next;
            } while (temp != head);

            if (nodeFound) {
                newNode.next = temp.next;
                newNode.prev = temp;
                temp.next.prev = newNode;
                temp.next = newNode;
                System.out.println("Node inserted in the middle successfully!");
            } else {
                System.out.println("Node with the specified value not found. Insertion failed.");
            }
        } else if (ch == 3) {
            newNode.next = head;
            newNode.prev = head.prev;
            head.prev.next = newNode;
            head.prev = newNode;
        }
    }

    public void Search() {
        Scanner sc = new Scanner(System.in);
        if (head == null) {
            System.out.println("\nList is empty");
            return;
        }
        System.out.println("Enter the value to search: ");
        int value = sc.nextInt();
        Node current = head;
        do {
            if (current.data == value) {
                System.out.println("Node with value " + value + " found.");
                return;
            }
            current = current.next;
        } while (current != head);
        System.out.println("Node with value " + value + " not found.");
    }

    public void Delete() {
        Scanner sc = new Scanner(System.in);
        if (head == null) {
            System.out.println("List is empty.");
            return;
        }
        System.out.println("Enter the value to delete: ");
        int value = sc.nextInt();
        Node current = head;
        do {
            if (current.data == value) {
                if (current == head) {
                    if (head.next == head) {
                        head = null;
                    } else {
                        head.prev.next = head.next;
                        head.next.prev = head.prev;
                        head = head.next;
                    }
                    System.out.println("Node with value " + value + " deleted successfully.");
                } else {
                    current.prev.next = current.next;
                    current.next.prev = current.prev;
                    System.out.println("Node with value " + value + " deleted successfully.");
                }
                return;
            }
            current = current.next;
        } while (current != head);
        System.out.println("Node with value " + value + " not found.");
    }

    public void Display() {
        if (head == null) {
            System.out.println("List is empty.");
            return;
        }
        Node current = head;
        System.out.print("- - - <-> ");
        do {
            System.out.print(current.data + " <-> ");
            current = current.next;
        } while (current != head);
        System.out.println("- - - ");
    }

    public void Count() {
        if (head == null) {
            System.out.println("List is empty.");
            return;
        }
        int count = 0;
        Node current = head;
        System.out.print("- - - <-> ");
        do {
            count++;
            current = current.next;
        } while (current != head);
        System.out.println("Number of nodes in the list: " + count);
    }

    public void Reverse() {
        if (head == null) {
            System.out.println("List is empty. Cannot reverse.");
            return;
        }

        Node current = head;
        Node temp;
        do {
            temp = current.prev;
            current.prev = current.next;
            current.next = temp;
            current = current.prev;
        } while (current != head);
        head = temp.prev;
        System.out.println("List is reversed successfully.");
    }

    public void Rotate() {
        Scanner sc = new Scanner(System.in);
        if (head == null) {
            System.out.println("List is empty. Cannot rotate.");
            return;
        }
        System.out.println("Enter the number of positions to rotate: ");
        int position = sc.nextInt();
        for (int i = 0; i < position; i++) {
            head = head.next;
        }
        System.out.println("List rotated by " + position + " positions.");
    }

    public static void main(String[] args) {
        CDL x = new CDL();
        int ch = 0;
        Scanner sc = new Scanner(System.in);

        do {
            System.out.println("\n1.Insert\n2.Search\n3.Delete\n4.Display\n5.Count\n6.Reverse\n7.Rotate\n8.Exit\n>>> ");
            ch = sc.nextInt();
            switch (ch) {
                case 1:
                    x.Insert();
                    break;
                case 2:
                    x.Search();
                    break;
                case 3:
                    x.Delete();
                    break;
                case 4:
                    x.Display();
                    break;
                case 5:
                    x.Count();
                    break;
                case 6:
                    x.Reverse();
                    break;
                case 7:
                    x.Rotate();
                    break;
            }
        } while (ch != 8);
    }
}

File: LinkedList/CircularDoublyLinkList/CDL.cpp
#include<iostream>
using namespace std;

class Node {
public:
    int data;
    Node* next;
    Node* prev;
};

class CDL {
public:
    Node* head;
    CDL() {
        head = nullptr;
    }

    void Insert() {
        cout << "\n1.Insert at Start \n2.Middle \n3.End >> ";
        int ch = 0, data = 0;
        cin >> ch;
        cout << "Enter the value >> ";
        cin >> data;
        Node* newNode = new Node();
        newNode->data = data;

        if (!head) {
            head = newNode;
            head->next = head;
            head->prev = head;
            return;
        }

        if (ch == 1) {
            newNode->next = head;
            newNode->prev = head->prev;
            head->prev->next = newNode;
            head->prev = newNode;
            head = newNode;
        } else if (ch == 2) {
            Node* temp = head;
            cout << "After which Node you want to insert data >> ";
            int nextNode = 0;
            cin >> nextNode;

            bool nodeFound = false;
            do {
                if (temp->data == nextNode) {
                    nodeFound = true;
                    break;
                }
                temp = temp->next;
            } while (temp != head);

            if (nodeFound) {
                newNode->next = temp->next;
                newNode->prev = temp;
                temp->next->prev = newNode;
                temp->next = newNode;
                cout << "Node inserted in the middle successfully!" << endl;
            } else {
                cout << "Node with the specified value not found. Insertion failed." << endl;
            }
        } else if (ch == 3) {
            newNode->next = head;
            newNode->prev = head->prev;
            head->prev->next = newNode;
            head->prev = newNode;
        }
    }

    void Search(int value) {
        if (head == nullptr) {
            cout<<"\nList is empty"<<endl;
            return;
        }
        Node* current = head;
        do {
            if (current->data == value) {
                cout<<"Node with value "<<value <<" found."<<endl;
                return;
            }
            current = current->next;
        } while (current != head);
        cout<<"Node with value "<<value<<" not found "<<endl;
    }

    void Delete(int value) {
        if (head == nullptr) {
            cout<<"List is empty."<<endl;
            return;
        }
        Node* current = head;
        do {
            if (current->data == value) {
                if (current == head) {
                    if (head->next == head) {
                        delete head;
                        head = nullptr;
                    } else {
                        head->prev->next = head->next;
                        head->next->prev = head->prev;
                        head = head->next;
                        delete current;
                    }
                } else {
                    current->prev->next = current->next;
                    current->next->prev = current->prev;
                    delete current;
                }
                cout << "Node with value " << value << " deleted successfully." << endl;
                return;
            }
            current = current->next;
        } while (current != head);
        cout << "Node with value " << value << " not found " << endl;
    }

    void Display() {
        if (head == nullptr) {
            cout<<"list is empty"<<endl;
            return;
        }
        Node* current = head;
        cout<<"- - - <-> ";
        do {
            cout<<current->data<<" <-> ";
            current = current->next;
        } while (current != head);
        cout<<"- - - "<<endl;
    }

    void Count() {
        if (head == nullptr) {
            cout<<"list is empty"<<endl;
            return;
        }
        int count = 0;
        Node* current = head;
        do {
            count++;
            current = current->next;
        } while (current != head);
        cout << "Number of nodes in the list: " << count << endl;
    }

    void Reverse() {
        if (head == nullptr) {
            cout << "List is empty. Cannot reverse." << endl;
            return;
        }

        Node* current = head;
        Node* temp = nullptr;
        do {
            temp = current->prev;
            current->prev = current->next;
            current->next = temp;
            current = current->prev;
        } while (current != head);
        head = temp->prev;
        cout<<"List is reversed successfully"<<endl;
    }

    void Rotate(int position) {
        if (head == nullptr) {
            cout << "List is empty. Cannot rotate." << endl;
            return;
        }
        for (int i = 0; i < position; i++) {
            head = head->next;
        }
        cout << "List rotated by " << position << " positions." << endl;
    }
};

int main() {

    CDL x;
    int ch = 0, value = 0;
    do {
        cout<<"\n1.Insert\n2.Search\n3.Delete\n4.Display\n5.Count\n6.Reverse\n7.Rotate\n8.Exit\n>>> ";
        cin>>ch;
        if (ch == 1) {
            x.Insert();
        } else if (ch == 2) {
            cout << "Enter the value to search: ";
            cin >> value;
            x.Search(value);
        } else if (ch == 3) {
            cout << "Enter the value to delete: ";
            cin >> value;
            x.Delete(value);
        } else if (ch == 4) {
            x.Display();
        } else if (ch == 5) {
            x.Count();
        } else if (ch == 6) {
            x.Reverse();
        } else if (ch == 7) {
            cout << "Enter the number of positions to rotate: ";
            cin >> value;
            x.Rotate(value);
        }
    } while (ch != 8);

    return 0;
}